core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java

```diff
@@ -1266,10 +1271,17 @@ private static RelDataType arrayInsertReturnType(SqlOperatorBinding opBinding) {
     // we don't need to do leastRestrictive on componentType and elementType,
     // because in operand checker we limit the elementType must equals array component type.
     // So we use componentType directly.
-    RelDataType type = componentType;
+    RelDataType type =
+        opBinding.getTypeFactory().leastRestrictive(
+            ImmutableList.of(componentType, elementType));
     if (elementType.isNullable()) {
       type = opBinding.getTypeFactory().createTypeWithNullability(type, true);
     }
+
+    if (componentType != elementType) {
```

Comment:
we can't use != to compare directly. try to use `equalsSansFieldNames`.

Reply:
done